# 3. Addressing Missing & Duplicate Values

Importing Libraries & Data

In [3]:
import pandas as pd
import numpy as np
import os

In [4]:
# Creating file path to get to Instacart Project file. 
path = r'C:\Users\Neena Tilton\Dropbox\Projects\01_2020_InstacartBasket'

In [5]:
# Importing products.csv dataframe from "OriginalData" folder.
# Importing wrangled data orders.csv from "PreparedData" folder.
df_products = pd.read_csv(os.path.join(path,'02_Data', 'OriginalData', 'products.csv'), index_col = False) 
df_orders = pd.read_csv(os.path.join(path,'02_Data', 'PreparedData', 'orders_wrangled.csv'), index_col = False) 

Finding and addressing missing values:

In [11]:
# To find missing values in dataframe.
df_products.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# To see only the null values, create a subset meeting condition of isnull() = True, and call for the new subset df. 
df_nan = df_products[df_products['product_name'].isnull() == True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
# Check the size of df before removing missing values.
df_products.shape

(49693, 5)

In [15]:
df_products_clean = df_products[df_products['product_name'].isnull() == False]

In [16]:
# Check size of new cleaned df to make sure exactly 16 rows are removed. 
df_products_clean.shape

(49677, 5)

Another way to drop missing values:

In [17]:
# This will overwrite the original dataframe and replace with new df with missing value rows dropped. 
df_products.dropna(subset = ['product_name'], inplace = True)

Finding and addressing duplicates:

In [18]:
# This will create a new subset containing of only rows that are duplicated.
df_dups = df_products_clean[df_products_clean.duplicated()]

In [19]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

Creating a new dataframe without the duplicates: 

In [20]:
# Check number of rows in df (this case it is that df_products_clean).
df_products_clean.shape

(49677, 5)

In [21]:
# Then, create a new dataframe that doesn’t include the duplicates identified above.
df_products_clean_no_dups = df_products_clean.drop_duplicates()

In [22]:
# Check the number of rows in the new df, confirm that only x number of rows identified earlier are removed, which should be
# 5 less than the original dataframe in this case. 
df_products_clean_no_dups.shape

(49672, 5)

Exporting to CSV after Addressing Missing & Duplicate Values

In [23]:
df_products_clean_no_dups.to_csv(os.path.join(path,'02_Data', 'PreparedData', 'products_checked.csv'))

Running descriptive analysis on the 'products' dataframe:

In [24]:
df_products.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [25]:
df_products_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

Things to take note: in product_id, there are 49672 enteries (or 49677 before dropping duplicates), but the max value is 49688, which means is that when we dropped the 16 rows with missing values for product_name, it created holes in the consecutive numbering of the ids. This is not really a problem, but important to note here. Other issue is that the max value for prices is $99,9999, this value is far greater than most other pricing, plus suspicious since we are dealing with groceries item. This price could be a typo, we need to look further into this. 

Checking for mixed-type data in the 'orders' dataframe:

In [27]:
# This for-loop finds mixed-type data in the dataframe.
for col in df_orders.columns.tolist():
    weird = (df_orders[[col]].applymap(type) != df_orders[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_orders[weird]) > 0:
        print(col)

Since there was no output, there was not a column with mixed-type data.

Running data type test to see what each column has as data type.:

In [30]:
df_orders.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_time_of_day           int64
days_since_prior_order    float64
dtype: object

Checking for missing values in the 'orders' dataframe: 

In [31]:
# Checking df_orders dataframe for missing observations. 
df_orders.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_time_of_day              0
days_since_prior_order    206209
dtype: int64

We have 206,209 missing values (of total 3,421,083 observations) in 'days_since_prior_order' column, which isn't odd since the first order for each new user will have 'NaN' recorded in this column. If anything, this number should match how many users (or total count of "user_id") there are. 

In [33]:
df_orders.shape

(3421083, 7)

In [37]:
# Running basic stats to see what max value user_id is.
df_orders.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_time_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Sure enough, the largest number for user_id is 206209; thus it makes sense that there are 206,209 "NaN" enteries under the 'days_since_prior_order' column. Since this is valuable information, the "NaN" enteries will not be removed or changed. 

Checking for duplicates in 'orders' dataframe: 

In [39]:
# Finding dupicates in df_orders dataframe and creating a subset of them.
df_orders_dups = df_orders[df_orders.duplicated()]

In [40]:
df_orders_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_time_of_day, days_since_prior_order]
Index: []

Since the new subset meant for duplicates has no rows, there were no duplicates. 

Exporting dataframe to csv:

In [41]:
df_orders.to_csv(os.path.join(path,'02_Data', 'PreparedData', 'orders_checked.csv'))